### PR TITLE
FM2-562: Fix pagination to address slow running queries

### PR DESCRIPTION
**Pull request by rkorytkowski**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Review by ibacher — APPROVED** [review]
Thanks @rkorytkowski! This looks great! My only real comment is that, for w/e reason, the OpenMRS coding standard is to never use `*` imports.

**Comment by rkorytkowski** [comment]
This is an initial pass. There's still one failing tests that I need to fix. I have trouble understanding the logic behind the method that is being tested so I'd appreciate some explanation.

The fix is to use DB pagination instead of paginating on the java side. Alongside I changed the code to run only 1 query and not two queries to get results in most cases as there's no need to first get IDs and then actual results. I also removed distinct filtering on results in places where it is not needed and made the distinct work on the DB side. The count also uses proper select count(*) and not size() on the results.

**Comment by rkorytkowski** [comment]
> Thanks @rkorytkowski! This looks great! My only real comment is that, for w/e reason, the OpenMRS coding standard is to never use `*` imports.

I'll fix that. Apparently my IDE is not configured properly.

**Comment by rkorytkowski** [comment]
@ibacher Thanks! Fixed the imports and the failing test. There's one thing with the new implementation for lastn obs feature that the count of results may be off in a rare condition. I commented in code. It feels such rare that I would not sacrifice performance to support it. I'm also not sure though if we expose that count anywhere...

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java [hunk 1]
```diff
@@ -59,26 +55,92 @@ public class FhirObservationDaoImpl extends BaseFhirDao<Obs> implements FhirObse
...
-			@SuppressWarnings("unchecked")
-			List<LastnResult<Integer>> results = ((List<Object[]>) criteria.list()).stream().map(objects -> {
-				Map<String, Object> attributes = new HashMap<>();
-				attributes.put("concept", objects[1]);
-				return new LastnResult<>((Integer) objects[0], (Date) objects[2], attributes);
-			}).collect(Collectors.toList());
+			List<Obs> results = new ArrayList<>();
+			int firstResult = 0;
+			int maxGroupCount = getMaxParameter(theParams);
+			final int BATCH_SIZE = 3;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java:69]: I'm not sure I quite understand why we're using the constant 3 here?

**Comment by ibacher** [comment]
@rkorytkowski The `count` does end up exposed as the `total` field in the search results. That said, I don't think that having that exactly right is super critical...

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java (hunk 1 above)
Comment (rkorytkowski) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirObservationDaoImpl.java:69]: I was testing the algorithm... I forgot to revert it back to 100. Good eye ;)

**Comment by rkorytkowski** [comment]
> @rkorytkowski The `count` does end up exposed as the `total` field in the search results. That said, I don't think that having that exactly right is super critical...

Actually, I misunderstood the algorithm a bit, which is why the integration tests were failing and it may happen more commonly. I made the adjustments to the count so it's always accurate.

I also noticed that getting counts was called twice during a request so decided to cache it for the lifespan of the request.

**Comment by ibacher** [comment]
> I also noticed that getting counts was called twice during a request so decided to cache it for the lifespan of the request.

Ooof... I hadn't that before. Thanks that should be good. I'm happy for you to merge this in whenever you're comfortable doing so, though it would be nice to hear back from the OHRI team whether this helps.

**Pull request merged**